optimade/server/config.py

```diff
@@ -37,29 +57,93 @@ class LogLevel(Enum):
 
 
 class SupportedBackend(Enum):
+    """Enumeration of supported database backends
+
+    - `elastic`: [Elasticsearch](https://www.elastic.co/).
+    - `mongodb`: [MongoDB](https://www.mongodb.com/).
+    - `mongomock`: Also MongoDB, but instead of using the
+        [`pymongo`](https://pymongo.readthedocs.io/) driver to connect to a live Mongo database
+        instance, this will use the [`mongomock`](https://github.com/mongomock/mongomock) driver,
+        creating an in-memory database, which is mainly used for testing.
+
+    """
 
     ELASTIC = "elastic"
     MONGODB = "mongodb"
     MONGOMOCK = "mongomock"
 
 
+def config_file_settings(settings: BaseSettings) -> Dict[str, Any]:
+    """Configuration file settings source.
+
+    Based on the example in the
+    [pydantic documentation](https://pydantic-docs.helpmanual.io/usage/settings/#adding-sources),
+    this function loads ServerConfig settings from a configuration file.
+
+    The file must be of either type JSON or YML/YAML.
+
+    Parameters:
+        settings: The `pydantic.BaseSettings` class using this function as a
+            `pydantic.SettingsSourceCallable`.
+
+    Returns:
+        Dictionary of settings as read from a file.
+
+    """
+    import json
+    import os
+    import yaml
+
+    encoding = settings.__config__.env_file_encoding
+    config_file = Path(os.getenv("OPTIMADE_CONFIG_FILE", DEFAULT_CONFIG_FILE_PATH))
+
+    res = {}
+    if config_file.is_file():
+        config_file_content = config_file.read_text(encoding=encoding)
+
+        try:
+            res = json.loads(config_file_content)
+        except json.JSONDecodeError as json_exc:
+            try:
+                # This can essentially also load JSON files, as JSON is a subset of YAML v1,
+                # but I suspect it is not as rigorous
+                res = yaml.safe_load(config_file_content)
+            except yaml.YAMLError as yaml_exc:
+                warnings.warn(
+                    f"Unable to parse config file {config_file} as JSON or YAML, using the "
+                    "default settings instead..\n"
+                    f"Errors:\n  JSON:\n{json_exc}.\n\n  YAML:\n{yaml_exc}"
+                )
+    else:
+        warnings.warn(
+            f"Unable to find config file at {config_file}, using the default settings instead."
+        )
+
+    if res is None:
+        # This can happen if the yaml loading doesn't succeed properly, e.g., if the file is empty.
```

Comment:
```suggestion
        # This can happen if the yaml loading doesn't succeed properly, e.g., if the file is empty.
        warnings.warn("Unable to load any settings from {config_file}, using the default settings instead.")
```